Intermittent test failure in master for :ballerina-cli:test

**Description:**
Noticed intermittent `:ballerina-cli:test` failure in master for the last couple of days. 

Experienced today as well for PR #29351 for 46ff43575043cc2331e4334bbd51d6a233516b69. (https://github.com/ballerina-platform/ballerina-lang/runs/2199038738)
But, syncing again with master and letting re-run the jobs get the build passed.

Failed again for #29248.

</details>
<details>
  <summary><b>log for the failing part</b>  (<i>click to expand</i>)</summary>

```java
commands-test-suite > command-tests > io.ballerina.cli.cmd.AddCommandTest > testAddCommandWithoutProject STANDARD_OUT
    ballerina: not a ballerina project (or any parent up to mount point)
    You should run this command inside a ballerina project


Running Tests

	winery

		[fail] testRunMain:

		    Failed!

		    	at ballerina.test.0_0_0:createBallerinaError(assert.bal:43)
		    	ballerina.test.0_0_0:assertTrue(assert.bal:53)
		    	foo.winery.0_1_0.tests.main_tests:testRunMain(tests/main_tests.bal:5)


		0 passing
		1 failing
		0 skipped


Running Tests

	winery

		No tests found


	winery.storage

		No tests found


Running Tests

	winery

		[pass] testRunMain

		1 passing
		0 failing
		0 skipped


Running Tests

	winery

		[pass] testRunMain

		1 passing
		0 failing
		0 skipped


Running Tests

	winery
Error: Exception in thread "main" java.io.FileNotFoundException: C:\Users\RUNNER~1\AppData\Local\Temp\b7a-cmd-test-4787296999007856051342127036168\build-test-resources\validProjectWithBuildOptions\target\cache\tests_cache\winery\module_status.json (Access is denied)

		[pass] testRunMain
	at java.base/java.io.FileOutputStream.open0(Native Method)

	at java.base/java.io.FileOutputStream.open(FileOutputStream.java:298)
		1 passing
	at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:237)
		0 failing
	at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:187)
		0 skipped
	at org.ballerinalang.test.runtime.Main.writeStatusToJsonFile(Main.java:133)

	at org.ballerinalang.test.runtime.Main.startTestSuit(Main.java:121)
	at org.ballerinalang.test.runtime.Main.main(Main.java:97)

commands-test-suite > command-tests > io.ballerina.cli.cmd.BuildCommandTest > testSingleFileWithBuildOptions FAILED
    java.lang.AssertionError: error: error occurred while running testsjava.nio.file.AccessDeniedException: C:\Users\RUNNER~1\AppData\Local\Temp\b7a-cmd-test-4787296999007856051342127036168\build-test-resources\validProjectWithBuildOptions\target\cache\tests_cache\winery\module_status.json
        at org.testng.Assert.fail(Assert.java:96)
        at io.ballerina.cli.cmd.BuildCommandTest.testSingleFileWithBuildOptions(BuildCommandTest.java:429)

Running Tests

	winery.storage

		[pass] beforeFunc

		1 passing
		0 failing
		0 skipped


	winery

		[pass] beforeFunc

		1 passing
		0 failing
		0 skipped


Running Tests

	winery

		[pass] testRunMain

		1 passing
		0 failing
		0 skipped
```		
</details>

--- a/cli/ballerina-cli/src/main/java/io/ballerina/cli/task/RunTestsTask.java
+++ b/cli/ballerina-cli/src/main/java/io/ballerina/cli/task/RunTestsTask.java
@@ -333,9 +333,11 @@ public class RunTestsTask implements Task {
         String json = gson.toJson(testReport).replaceAll("\\\\\\(", "(");
 
         File jsonFile = new File(reportDir.resolve(RESULTS_JSON_FILE).toString());
-        try (Writer writer = new OutputStreamWriter(new FileOutputStream(jsonFile), StandardCharsets.UTF_8)) {
-            writer.write(new String(json.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
-            out.println("\t" + jsonFile.getAbsolutePath() + "\n");
+        try (FileOutputStream fileOutputStream = new FileOutputStream(jsonFile)) {
+            try (Writer writer = new OutputStreamWriter(fileOutputStream, StandardCharsets.UTF_8)) {
+                writer.write(new String(json.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
+                out.println("\t" + jsonFile.getAbsolutePath() + "\n");
+            }
         }
 
         Path reportZipPath = Paths.get(System.getProperty(BALLERINA_HOME)).resolve(BALLERINA_HOME_LIB).
@@ -344,18 +346,22 @@ public class RunTestsTask implements Task {
         if (Files.exists(reportZipPath)) {
             String content;
             try {
-                CodeCoverageUtils.unzipReportResources(new FileInputStream(reportZipPath.toFile()),
-                        reportDir.toFile());
+                try (FileInputStream fileInputStream = new FileInputStream(reportZipPath.toFile())) {
+                    CodeCoverageUtils.unzipReportResources(fileInputStream,
+                            reportDir.toFile());
+                }
                 content = Files.readString(reportDir.resolve(RESULTS_HTML_FILE));
                 content = content.replace(REPORT_DATA_PLACEHOLDER, json);
             } catch (IOException e) {
                 throw createLauncherException("error occurred while preparing test report: " + e.toString());
             }
             File htmlFile = new File(reportDir.resolve(RESULTS_HTML_FILE).toString());
-            try (Writer writer = new OutputStreamWriter(new FileOutputStream(htmlFile), StandardCharsets.UTF_8)) {
-                writer.write(new String(content.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
-                out.println("\tView the test report at: " +
-                        FILE_PROTOCOL + Paths.get(htmlFile.getPath()).toAbsolutePath().normalize().toString());
+            try (FileOutputStream fileOutputStream = new FileOutputStream(htmlFile)) {
+                try (Writer writer = new OutputStreamWriter(fileOutputStream, StandardCharsets.UTF_8)) {
+                    writer.write(new String(content.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
+                    out.println("\tView the test report at: " +
+                            FILE_PROTOCOL + Paths.get(htmlFile.getPath()).toAbsolutePath().normalize().toString());
+                }
             }
         } else {
             String reportToolsPath = "<" + BALLERINA_HOME + ">" + File.separator + BALLERINA_HOME_LIB +
@@ -473,10 +479,14 @@ public class RunTestsTask implements Task {
 
         Path jsonFilePath = Paths.get(testsCachePath.toString(), TesterinaConstants.TESTERINA_TEST_SUITE);
         File jsonFile = new File(jsonFilePath.toString());
-        try (Writer writer = new OutputStreamWriter(new FileOutputStream(jsonFile), StandardCharsets.UTF_8)) {
-            Gson gson = new Gson();
-            String json = gson.toJson(testSuiteMap);
-            writer.write(new String(json.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
+        try (FileOutputStream fileOutputStream = new FileOutputStream(jsonFile)) {
+            try (Writer writer = new OutputStreamWriter(fileOutputStream, StandardCharsets.UTF_8)) {
+                Gson gson = new Gson();
+                String json = gson.toJson(testSuiteMap);
+                writer.write(new String(json.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
+            } catch (IOException e) {
+                throw LauncherUtils.createLauncherException("couldn't write data to test suite file : " + e.toString());
+            }
         } catch (IOException e) {
             throw LauncherUtils.createLauncherException("couldn't write data to test suite file : " + e.toString());
         }

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
@@ -757,10 +757,15 @@ public class BTestRunner {
     private void writeFailedTestsToJson(List<String> failedTests, File jsonFile) {
         String errorMsg;
 
-        try (Writer writer = new OutputStreamWriter(new FileOutputStream(jsonFile), StandardCharsets.UTF_8)) {
-            Gson gson = new Gson();
-            String json = gson.toJson(failedTests);
-            writer.write(new String(json.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
+        try (FileOutputStream fileOutputStream = new FileOutputStream(jsonFile)) {
+            try (Writer writer = new OutputStreamWriter(fileOutputStream, StandardCharsets.UTF_8)) {
+                Gson gson = new Gson();
+                String json = gson.toJson(failedTests);
+                writer.write(new String(json.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
+            } catch (IOException e) {
+                errorMsg = "Could not write to Rerun Test json. Rerunning tests will not work";
+                errStream.println(errorMsg + ":" + e.getMessage());
+            }
         } catch (IOException e) {
             errorMsg = "Could not write to Rerun Test json. Rerunning tests will not work";
             errStream.println(errorMsg + ":" + e.getMessage());

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/Main.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/Main.java
@@ -130,10 +130,12 @@ public class Main {
         if (!Files.exists(tmpJsonPath.getParent())) {
             Files.createDirectories(tmpJsonPath.getParent());
         }
-        try (Writer writer = new OutputStreamWriter(new FileOutputStream(jsonFile), StandardCharsets.UTF_8)) {
-            Gson gson = new Gson();
-            String json = gson.toJson(moduleStatus);
-            writer.write(new String(json.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
+        try (FileOutputStream fileOutputStream = new FileOutputStream(jsonFile)) {
+            try (Writer writer = new OutputStreamWriter(fileOutputStream, StandardCharsets.UTF_8)) {
+                Gson gson = new Gson();
+                String json = gson.toJson(moduleStatus);
+                writer.write(new String(json.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
+            }
         }
     }
 

